### ISSUE TITLE
Support v0.8.4 custom errors

### ISSUE THREAD

**Issue by paulrberg** [protofire/solhint#289]
See the announcement [article](https://blog.soliditylang.org/2021/04/21/custom-errors/) for the new custom error syntax.

For the following code:

```solidity
/// Too big value for ceiling.
/// @param x Passed input.
error Ud60x18CeilOverflow(uint256 x);
```

I get this error in VSCode:

> Linter: Parse error: missing 'constant' at 'Ud60x18CeilOverflow' [undefined]

**Comment by fvictorio** [comment]
Hey Paul, I just published `solhint@3.3.5` which uses the latest version of the solidity parser and shouldn't throw this error. Please let me know if it works for you.

**Comment by paulrberg** [comment]
Nice, thanks Franco! Unfortunately it still doesn't work well in all cases. Take the following code:

```solidity
error SomeError();

contract Greeter {
    function throwError() external pure {
        revert SomeError();
    }
}

Yields the following error:

```
Linter: Parse error: mismatched input '(' expecting {';', '='} [undefined]
```

**Comment by fvictorio** [comment]
@paulrberg are you sure? I just tested that snippet with the latest published solhint version and I didn't get a parser error.

**Comment by paulrberg** [comment]
Yup, I just re-tested now, and I double-checked the version of `solhint` that I have installed, and it's 3.3.5. Furthermore, I installed the `@solidity-parser/parser` dependency manually, in case the versions that are depended on by other dependencies interfered somehow. That didn't work either.

Steps to reproduce

1. git clone git@github.com:paulrberg/solidity-template.git
2. yarn add solhint@^3.3.5 --dev
3. Add the error definition above
4. Add the `throwError` function above

**Comment by fvictorio** [comment]
Ugh, I have no idea what's going on here, but removing `yarn.lock` and re-installing fixes it. Do you know what could be the reason?

**Comment by paulrberg** [comment]
Haha there is something funky going on here! I deleted `yarn.lock` and re-installed the dependencies, but I am still getting the same error in VSCode.

I am using yarn@2.4.1, and you should be using the same, due to the `.yarnrc.yml` file which forces the use of yarn berry. But maybe you're using an older version of the global yarn CLI which doesn't recognise the berry config files? Could you run `yarn --version` outside the "solidity-template" folder?

Alternatively, maybe this is caused by [vscode-solidity](https://github.com/juanfranblanco/vscode-solidity)? This is unlikely, because of the "Linter: " prefix, but possible.

**Comment by fvictorio** [comment]
I'm pretty sure I'm using yarn v1.

Maybe the problem is that I used an RC version of the parser as the dependency, because I'm an idiot (I immediately published the non-RC version of the parser, which is exactly the same). I'm going to add a fix and release a new version of solhint soon, so I'll use that as an opportunity to also change the parser version. I'll let you know so you can try again with that change.

**Comment by paulrberg** [comment]
Yarn v1 activates v2 whenever it finds a `.yarnrc.yaml` file and a `.yarn` folder.

Haha ok.

**Comment by fvictorio** [comment]
@paulrberg can you try the latest published version? It doesn't have an RC as a dependency anymore. Maybe that was the issue.

**Comment by paulrberg** [comment]
Just tried now, unfortunately it still didn't work. I created a node.js project from scratch and I installed only `solhint` in it, in order to exclude any potential issues re dependency injections. Maybe this is caused by vscode-solidity? This is my user configuration:

```json
"solidity.linter": "solhint",
"solidity.compileUsingRemoteVersion": "v0.8.4+commit.c7e474f2",
"solidity.defaultCompiler": "remote"
```

The version of the extension that I'm using is 0.0.114.

**Comment by fvictorio** [comment]
Oh yes, it's entirely possible that the vscode extension is the problem. Can you try it directly in the terminal?

**Comment by paulrberg** [comment]
Oh, but I knew that in the command-line the linting works. Sorry, I should have made that clear. In the terminal it works since v3.3.5.

I opened an issue in the vscode-solidity repo, the issue must be there.

**Comment by fvictorio** [comment]
Ohh, sorry, I totally missed the "VSCode" part in the original issue :sweat_smile: Thanks for letting me know!
